concurrent/src/main/java/io/airlift/concurrent/ResourceLoader.java

```
@@ -0,0 +1,224 @@
+        }
+        catch (Exception e) {
+            exceptionHandler.accept(e);
+        }
+    }
+
+    public interface FileLoader<T>
+    {
+        T loadFile(File file)
+                throws Exception;
+    }
+
+    public interface SimpleRecurringTaskRunner
+    {
+        Runnable scheduleTask(Runnable runnable, Duration fixedDelay);
+    }
+
+    private static class InternalScheduledRecurringTaskRunner
+            implements SimpleRecurringTaskRunner
+    {
+        private final String name;
+
+        private InternalScheduledRecurringTaskRunner(String name)
+        {
+            this.name = requireNonNull(name, "name is null");
+        }
+
+        @Override
+        public Runnable scheduleTask(Runnable runnable, Duration fixedDelay)
+        {
+            ScheduledExecutorService executor = newSingleThreadScheduledExecutor(daemonThreadsNamed(name + "-background-loader"));
+            long fixedDelayNanos = fixedDelay.roundTo(NANOSECONDS);
+            executor.scheduleAtFixedRate(runnable, fixedDelayNanos, fixedDelayNanos, NANOSECONDS);
+            return executor::shutdownNow;
+        }
+    }
+
+    private static class NoOpScheduledRecurringTaskRunner
+            implements SimpleRecurringTaskRunner
+    {
+        @Override
+        public Runnable scheduleTask(Runnable runnable, Duration fixedDelay)
+        {
+            return () -> {};
+        }
+    }
+
+    public static <T> ResourceLoaderBuilder<T> builder(FileLoader<T> fileLoader)
+    {
+        return new ResourceLoaderBuilder<>(fileLoader);
+    }
+
+    public static final class ResourceLoaderBuilder<T>
+    {
+        private final FileLoader<T> fileLoader;
+        private String name;
+        private File file;
+        private Duration refreshPeriod = new Duration(0, NANOSECONDS);
+        private SimpleRecurringTaskRunner recurringTaskRunner;
+        private Consumer<Exception> exceptionHandler = exception -> log.error("Error refreshing " + name, exception);
```

I think by default we should propagate the failure to user.